docs/codeql/codeql-language-guides/customizing-library-models-for-javascript.rst

```diff
@@ -0,0 +1,468 @@
+
+.. code-block:: js
+
+  window.addEventListener("onclick", function (event) {
+    let data = event.data; // <-- 'event.data' became a spurious taint source
+  });
+
+
+We can refine the model by adding the **WithStringArgument** component to restrict the set of calls being considered:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/javascript-all
+        extensible: sourceModel
+      data:
+        - [
+            "global",
+            "Member[addEventListener].WithStringArgument[0=message].Argument[1].Parameter[0].Member[data]",
+            "remote",
+          ]
+
+The **WithStringArgument[0=message]** component here selects the subset of calls to **addEventListener** where the first argument is a string literal with the value **"message"**.
+
+Example: Using types to add MySQL injection sinks
+-------------------------------------------------
+
+In this example, we'll show how to add the following SQL injection sink:
+
+.. code-block:: ts
+
+  import { Connection } from "mysql";
+
+  function submit(connection: Connection, q: string) {
+    connection.query(q); // <-- add 'q' as a SQL injection sink
+  }
+
+We can recognize this using the following extension:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/javascript-all
+        extensible: sinkModel
+      data:
+        - ["mysql.Connection", "Member[query].Argument[0]", "sql-injection"]
+
+
+- The first column, **"mysql.Connection"**, begins the search at any expression whose value is known to be an instance of
+  the **Connection** type from the **mysql** package. This will select the **connection** parameter above because of its type annotation.
+- **Member[query]** selects the **query** member from the connection object.
+- **Argument[0]** selects the first argument of a call to that member.
+- **sql-injection** indicates that this is considered a sink for the SQL injection query.
+
+This works in this example because the **connection** parameter has a type annotation that matches what the model is looking for.
+
+In the next section, we'll show how to generalize the model to handle the absence of type annotations.
+
+Continued example: Dealing with untyped code
+--------------------------------------------
+
+Suppose we want the model from above to detect the sink in this snippet:
+
+.. code-block:: js
+
+  import { getConnection } from "@example/db";
+  let connection = getConnection();
+  connection.query(q); // <-- add 'q' as a SQL injection sink
+
+There is no type annotation on **connection**, and there is no indication of what **getConnection()** returns.
+Using a **typeModel** tuple we can tell our model that this function returns an instance of **mysql.Connection**:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/javascript-all
+        extensible: typeModel
+      data:
+        - ["mysql.Connection", "@example/db", "Member[getConnection].ReturnValue"]
```

Comment:
Entirely optional, but do you think that adding some column names here would be helpful?

```suggestion
         # Type,               Import,        Path         
        - ["mysql.Connection", "@example/db", "Member[getConnection].ReturnValue"]
```

Doing something similar above might be nice too.